router/tree.go

```diff
@@ -314,68 +312,60 @@ func (t *spanningTree) Update(p *Peer, newUpdate types.SwitchAnnouncement) error
 	t.updateMutex.Lock()
 	defer t.updateMutex.Unlock()
 
-	lastGlobalUpdate := t.Root()
-	globalKeyDelta := newUpdate.RootPublicKey.CompareTo(lastGlobalUpdate.RootPublicKey)
-	globalTimeSince := time.Since(lastGlobalUpdate.at)
-	globalUpdate := false
+	if !globalUpdate {
+		lastGlobalUpdate := t.Root()
```

Comment:
Since the cached last root announcement has been removed from the tree, in the case that this new announcement came from our parent, the value returned from `Root` will have already been overwritten with the new announcement at this point from this line above:
https://github.com/matrix-org/pinecone/blob/d620f99b32e9bf9dd48e2de8735eedf15aeee212/router/tree.go#L302

To keep with removing the cached root announcement, the root announcement could be copied right before the call to `updateAnnouncement` using `lastRootAnnouncement := *t.Root()` and use this copy of the old announcement for the lastGlobalUpdate and oldRoot information:
``` suggestion
		lastGlobalUpdate := &lastRootAnnouncement
```

And this line:
https://github.com/matrix-org/pinecone/blob/d620f99b32e9bf9dd48e2de8735eedf15aeee212/router/tree.go#L355

To:
``` go
	oldcoords, oldroot := lastRootAnnouncement.Coords(), lastRootAnnouncement.RootPublicKey
```

Reply:
Yes, indeed (although I don't think we need to make specific pointerised copies, because the update already replaces the entire pointer value anyway). I still can't entirely decide if caching the last announcement is better or worse than just generating one from the last peer announcement, like this PR does.

Reply:
Agreed. I'm not sure which approach is better.
I kind of like not having the extra field in the spanningTree struct. The information is only valuable during the time of update. So removing it from the struct simplifies the data structure a little bit and makes it less complex to reason about.

But other than that, I don't have a strong opinion either way.